### PR TITLE
ITHC recommended headers

### PR DESCRIPTION
**Pull request by spatel033**
ITHC recommended to add and update headers. Some headers were deprecated and incompatible across all browsers. This PR to fix security headers, report was also generated and shared on [this card](https://trello.com/c/EfkvM0wg/1070-ensure-recommended-headers-are-applied-for-admin-app).

There are headers updated related to font, style and script, not sure if this would break anything on server.

image
image

### CODE REVIEW

**Comment by kr8n3r** [comment]
quick glance: this is at minimum breaking JS execution.  we need to get this onto a dev environment to test in something close to production

File: app/__init__.py [hunk 1]
```diff
@@ -370,6 +371,16 @@ def useful_headers_after_request(response):
     response.headers.add("Cache-Control", "no-store, no-cache, private, must-revalidate")
     for key, value in response.headers:
         response.headers[key] = SanitiseASCII.encode(value)
+    response.headers.add("Strict-Transport-Security", "max-age=31536000; preload")
+    response.headers.add("Referrer-Policy", "strict-origin-when-cross-origin")
```
Comment (kr8n3r) [inline app/__init__.py:375]: this is fine

File: app/__init__.py [hunk 2]
```diff
@@ -370,6 +371,16 @@ def useful_headers_after_request(response):
...
+    response.headers.add("Cross-Origin-Embedder-Policy", "require-corp")
+    response.headers.add("Cross-Origin-Opener-Policy", "same-origin")
+    response.headers.add("Cross-Origin-Resource-Policy", "same-origin")
+    response.headers.add(
+        "Permissions-Policy",
+        "geolocation=(), microphone=(), camera=(), autoplay=(), payment=(), sync-xhr=()",
```
Comment (kr8n3r) [inline app/__init__.py:381]: fine, to be there as we don't use most of this, though we might need to check synchronous xhr requests

File: app/__init__.py [hunk 3]
```diff
@@ -338,19 +338,20 @@ def load_user_id_before_request():
...
-    response.headers.add("X-Frame-Options", "deny")
     response.headers.add("X-Content-Type-Options", "nosniff")
     response.headers.add("X-XSS-Protection", "1; mode=block")
     response.headers.add(
         "Content-Security-Policy",
         (
-            "default-src 'self' {asset_domain} 'unsafe-inline';"
-            "script-src 'self' {asset_domain} 'unsafe-inline' 'unsafe-eval' data:;"
+            "default-src 'self' {asset_domain};"
+            "script-src 'self' {asset_domain};"
```
Comment (kr8n3r) [inline app/__init__.py:347]: this breaks JS as a SHA will need to be created / added

**Comment by kr8n3r** [comment]
just the JS errors. some things we can sort out easily, other's not as it's legacy JS
Screenshot 2024-12-05 at 10 37 28

File: app/__init__.py [hunk 4]
```diff
@@ -370,6 +371,16 @@ def useful_headers_after_request(response):
     response.headers.add("Cache-Control", "no-store, no-cache, private, must-revalidate")
     for key, value in response.headers:
         response.headers[key] = SanitiseASCII.encode(value)
+    response.headers.add("Strict-Transport-Security", "max-age=31536000; preload")
+    response.headers.add("Referrer-Policy", "strict-origin-when-cross-origin")
+    response.headers.add("Cross-Origin-Embedder-Policy", "require-corp")
```
Comment (kr8n3r) [inline app/__init__.py:376]: Given that COEP could block assets from loading, it's a good idea to leverage the Report Only version of the header coupled with reporting via the Reporting API first to ensure that no harm will be caused to your site.

`Cross-Origin-Embedder-Policy-Report-Only: (unsafe-none|require-corp); report-to="default"`

we need to carefully check in a prod-like environemt

File: app/__init__.py [hunk 5]
```diff
@@ -370,6 +371,16 @@ def useful_headers_after_request(response):
...
+    response.headers.add("Cross-Origin-Opener-Policy", "same-origin")
+    response.headers.add("Cross-Origin-Resource-Policy", "same-origin")
```
Comment (kr8n3r) [inline app/__init__.py:378]: this will likely cause static logos and static asstes on prod not to load as they're on a different origin to the admin app

**Comment by kr8n3r** [comment]
_Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self'". Either the 'unsafe-inline' keyword, a hash ('sha256-Nv8QC5KMerDUPUkQ13N3EPtBelVQH+r4Wo3DV9ADatk='), or a nonce ('nonce-...') is required to enable inline execution._

this is for 

```
  <style>
      .govuk-header__container { border-color: #81878b }
  </style>
```

I wonder if we need this @tombye ?

**Comment by quis** [comment]
> I wonder if we need this @tombye ?

It’s still really useful to have a different colour for each environment. We could move this into a stylesheet served from the admin app though.

File: app/__init__.py (hunk 3 above)
Comment (kr8n3r) [reply app/__init__.py]: i've added a SHA for govuk-frontend

**Comment by kr8n3r** [comment]
> > I wonder if we need this @tombye ?
> 
> It’s still really useful to have a different colour for each environment. We could move this into a stylesheet served from the admin app though.

yeah something like that  @quis. quick search shows that inline styles are used in a bunch of places; letter preview and svgs, so need to check if this `style-src` should remain as is

**Comment by tombye** [comment]
I've chatted with @leohemsted and @spatel033 about the points raised by @kr8n3r. We think we need to not block this PR with trying to solve the problems with:
- permissions policy (`sync-xhr`)
- Content Security Policy `insafe-inline` for scripts and CSS, `unsafe-eval` and `data:`

I'll write up a card to spike solutions to these so they don't block this PR.

**Comment by kr8n3r** [comment]
we need to remove that directive or change the way we load the branding so it's not in an iframe
Screenshot 2024-12-06 at 10 07 43

**Review by kr8n3r — APPROVED** [review]
I've tested this reduced scope changes as much as possible on local and dev-a. Things appear to work as normal , but we should double check after deploying, first on preview and then on prod

**Pull request merged**